Fixes for wildcard/dual certificates and non-dynamic domains, plus cleanup hook script.

PharmaJap opened the pull request:
Included patches:

- Spacing and formatting fixups for clarity, and as recommended by `shellcheck`.
- Echo status and success messages to stdout instead of stderr.
- Strips leading `*.`s from the domain name to enable verification of wildcard certificates.
- Reads and republishes existing `_acme-challenge`s along with the current challenge, to enable verification of wildcard certificates with SubjectAlternativeNames (e.g. one certificate for *.foo.bar and foo.bar).
- Reads and respects the domain's `minimum_ttl`. This allows the script to be used on any domain served by deSEC, instead of failing on non-`dedyn.io` domains.
- Adds cleanup functionality when called with `--manual-cleanup-hook`.
- Adds documentation of cleanup script to `README.md`

Closes #7 and #12.

peterthomassen commented on hook.sh:
```diff
@@ -39,7 +39,7 @@ if [ ! "$(type -P curl)" ]; then
 fi
 
-echo "Setting challenge to ${CERTBOT_VALIDATION} ..."
+echo -e "\e[32mSetting challenge to ${CERTBOT_VALIDATION} ...\e[0m"
```
What is the purpose of this? Usually, color is used for errors or success messages, but not for regular progress printing. (certbot itself for example also does not use green color for their standard output.)

I would propose to keep stdout without color.

PharmaJap replied:
Simply for visibility, to separate the hook's messages from certbot's. I'm not attached to it.

With the redirect to stderr (`>&2`), messages were being caught by certbot and not displayed until after the hook exited, in red, with an "error output" header. So the user would get a long delay with no output, and then a confusing dump of success messages that look like errors. Messages sent to stdout display in real time.

I can remove the escape codes, but I think we should still send status and success messages to stdout instead of stderr. Agreed?

peterthomassen replied:
Yes, fully! My remark was only about the color :-)

peterthomassen replied:
Also, I'm not attached to removing the coloring. I just wondered about the reason for it, because it looks like a success message before there is actual success.

peterthomassen commented on hook.sh:
```diff
@@ -5,79 +5,102 @@
->&2 echo "Verifying challenge is set correctly. This can take up to 2 minutes."
->&2 echo "Current Time: `date`"
+readarray -t acme_live <<< "$(curl "${args[@]}" -X GET "https://desec.io/api/v1/domains/$DEDYN_NAME/rrsets/_acme-challenge$infix/TXT" \
+    | grep -o '"records":\[[^]]*]' | grep -o '\\"[^,]*\\"')"
```
I have several comments here:

- The URL is missing the final slash. This causes an unnecessary redirect.
- You can use `readarray -t acme_live < <(curl ...)` to simplify and avoid the unnecessary string parsing.
- The code make assumptions about the JSON formatting. Arbitrary whitespace is allowed before and and after the colon `:`. To take this into account, I suggest piping the `curl` output through `tr -d '\n'` and replacing the `:` in the regex by `[[:space:]]*:[[:space:]]*`.
- Records are currently split by `,` and later reassembled. First of all, this breaks in case a challenge contains a comma `,`. Second, I don't think it is necessary at all to split the records. You can just `grep -o '"\".*\""'` to obtain the whole list of JSON strings (without the enclosing brackets). Later, you can just add the current challenge if it is not contained in this string.
- Also, this command writes `curl: (22) The requested URL returned error: 404` to stderr if the RRset does not yet exist, so the user will be confronted with an error even though everything is fine. It would be better to handle this case gracefully. The easiest way to achieve this may be to query `rrsets/$DEDYN_NAME/rrsets/?subname=_acme-challenge$infix&type=TXT`, which returns a JSON array with zero or one elements. If the RRset is not found, the return value will be `[]`, but the exit code will be a success to. The regular expression should work nevertheless.

PharmaJap replied:
Great ideas all around! Reworking this to read the whole records array into an ordinary variable and appending the challenge (if it doesn't already exist) saves us an array and a loop.

Great to know about the final slash/redirect. So many future headaches avoided :-)

peterthomassen commented on hook.sh:
```diff
@@ -46,17 +46,40 @@ echo -e "\e[32mSetting challenge to ${CERTBOT_VALIDATION} ...\e[0m"
+readarray -t acme_live <<< "$(curl "${args[@]}" -X GET "https://desec.io/api/v1/domains/$DEDYN_NAME/rrsets/_acme-challenge$infix/TXT" \
+    | grep -o '"records":\[[^]]*]' | grep -o '\\"[^,]*\\"')"
+
+[ ${#acme_live[@]} -gt 0 ] && [ -n "${acme_live[*]}" ] && [ -n "$timeout_ttl" ] && \
```
- Is it even possible that the first and the second condition are not true at the same time?
- The last condition is also true if you set `timeout_ttl=no`. The comment further up implies that `sleep` is only called if the values is `yes`. We should either be checking that here, or change the comment to say `Set timeout_ttl to any value ...`.

PharmaJap replied:
Yes, but only because a null string in an array counts as an entry. Will be avoided by the changes above.

I agree with the timeout ambiguity. I'm going to make the comments a little more general, and the comparison more explicit.

peterthomassen commented on hook.sh:
```diff
@@ -46,17 +46,40 @@ echo -e "\e[32mSetting challenge to ${CERTBOT_VALIDATION} ...\e[0m"
+# certificates for the same domain (e.g. RSA and ECDSA), or both, we'll need to
+# wait 60 seconds for the records to expire before verification begins. We'll
+# store any existing challenges and publish them along with the current one.
+# Set timeout_ttl=yes in ./dedynauth if you want to enable this.
```
The last word (`this`) is a bit unclear. I know it refers to the 60 second wait, but a reader may think it is about wildcards or multi-algorithm use.

PharmaJap replied:
As above.

peterthomassen commented on hook.sh:
```diff
@@ -46,17 +46,40 @@ echo -e "\e[32mSetting challenge to ${CERTBOT_VALIDATION} ...\e[0m"
-    '-o' '/dev/null' \
 )
 
+# Letsencrypt uses a TTL of 60 seconds on their resolvers. If we're getting
```
Does LE really use a 60 seconds TTL on their servers? My hypothesis is that they always query the authoritative servers, so that the TTL does not play any role, and the only relevant timing is the replication time for the DNS slaves (but I'm not sure about that). Do you have a reference for your statement?

PharmaJap replied:
The mechanics aren't super well-documented, but I was going off of this comment from one of the LE engineers: https://community.letsencrypt.org/t/dns-servers-used-by-letsencrypt-for-challenges/32127/10

In my test case, publishing 2 challenges for an ECDSA certificate (and verifying they were available before returning to certbot) succeeded, but immediately afterwards publishing 2 challenges (for the same domain) for an RSA certificate would fail, even if I verified i could `dig` those records `@ns1.desec.io`. I was assuming the old records were still stuck in the Unbound cache cpu mentioned.

Any idea what the typical propagation time is between the slaves? I can play with this value somewhat and save people the wait.

peterthomassen replied:
The propagation time between our API responding with a success code and the slaves actually serving that content is usually a few seconds, at least in Europe. In a few percent of cases, however, it can take up to around one minute.

We spent a lot of effort on getting it really fast and reliable, but it's actually a very tricky issue. We hope to improve it further in the future ...

peterthomassen commented on hook.sh:
```diff
@@ -46,17 +46,40 @@ echo -e "\e[32mSetting challenge to ${CERTBOT_VALIDATION} ...\e[0m"
+[ ${#acme_live[@]} -gt 0 ] && [ -n "${acme_live[*]}" ] && [ -n "$timeout_ttl" ] && \
+echo -e "\e[32mWaiting 60 seconds for Letsencrypt resolvers to clear ...\e[0m" &&  sleep 60
+
+for ((i = 0; i < "${#acme_live[@]}"; i++));do
```
If we don't split the TXT records further up, we don't need the loop here. We can just check whether `current_value='"\"'"$CERTBOT_VALIDATION"'\""'` is a substring of the records string, and if it's not the case, add `,${current_value}` to the string.

PharmaJap replied:
As above :-)

peterthomassen commented on hook.sh:
```diff
@@ -77,9 +77,13 @@ for ((i = 0; i < "${#acme_live[@]}"; i++));do
 
+# Find minimum_ttl for the domain and use that instead of hardcoding a ttl.
+# This allows use on any deSEC domain (Letsencrypt ignores TTL > 60).
+minimum_ttl=$(curl "${args[@]}" -X GET "https://desec.io/api/v1/domains/$DEDYN_NAME/" | grep -o "\"minimum_ttl\":[0-9]*" | grep -o "[0-9]*")
```
- Again, there are assumptions about JSON formatting here. To work around that, we can remove newlines from the `curl` output and then allow arbitrary whitespace before and after the colon (see my other content).
- If we replace the outside quotes with single quotes, the inner ones don't need escaping.

PharmaJap replied:
Good catch! Thanks.

peterthomassen commented on cleanup.sh:
```diff
@@ -0,0 +1,67 @@
+
+args=( \
+    '-sSLf' \
+    '--post301' \
```
This line is unnecessary when adding a final slash to the URL.

PharmaJap replied:
So many future headaches saved!

peterthomassen commented on cleanup.sh:
```diff
@@ -0,0 +1,67 @@
+    '-o' '/dev/null' \
+    '-H' "Authorization: Token $DEDYN_TOKEN" \
+    '-H' 'Accept: application/json' \
+    '-H' 'Content-Type: application/json' \
```
These lines are unnecessary for `DELETE` requests.

PharmaJap replied:
Fair enough. A little too happy with the copy/paste on this one :-)

peterthomassen commented on hook.sh:
```diff
@@ -46,17 +46,40 @@ echo -e "\e[32mSetting challenge to ${CERTBOT_VALIDATION} ...\e[0m"
 
+# Remove leading wildcard from infix, if present
+# *.foobar.dedyn.io gives "" while *.a.foobar.dedyn.io gives ".a"
+infix=${infix#.\*}
```
Cool. Good catch!

peterthomassen replied:
However, should be `infix=${infix#\*.}`

PharmaJap replied:
Am I overlooking something obvious? The code in main adds a leading dot, then strips off a trailing dot and the `DEDYN_NAME`, so that the infix can be appended to `_certbot-challenge` directly. If there happens to be a wildcard, we want to strip off the first dot and the wildcard, but still leave the infix with a leading dot:

```
CERTBOT_DOMAIN=*.a.foobar.dedyn.io
DEDYN_NAME=foobar.dedyn.io
domain=.$CERTBOT_DOMAIN			# .*.a.foobar.dedyn.io
infix=${domain%.$DEDYN_NAME}		# .*.a
infix=${infix#.\*}			# .a
```

peterthomassen replied:
Ah. Of course, that's right. I had overlooked the fact that we add a dot in the beginning. Sorry!

peterthomassen reviewed (changes requested):
First of all, thanks so much for opening this PR! It's great to receive contributions like this to improve our script. Thank you for all the work!

I have a few comments to simplify the script and make it more resilient against error conditions and user misunderstandings. It would be great if you could take a look! :-)

nils-wisiol commented:
@PharmaJap thanks for your contribution! I really appreciate it.

Just for the legal side of things, could you please state the license under which your contribution will be licenced? Note that we can only merge MIT-compatible licenses, and I'd prefer to avoid mixed licencing.

PharmaJap commented:
@nils-wisiol MIT should be fine; I'll work it in with the changes I'm working on for Peter.

Is slipping it into the commit messages enough? Or should it be stated explicitly somewhere?

nils-wisiol commented:
A definite statement in this PR conversation is fine. Let's not make things too complicated and hope the best :-)

PharmaJap committed:
style(hook): clean up formatting, whitespace, and variable declarations

Copyright (c) 2020 Justin Aplin for deSEC
Licensed under the MIT License as specified in LICENSE

PharmaJap committed:
fix(hook): echo non-error messages to stdout

Copyright (c) 2020 Justin Aplin for deSEC
Licensed under the MIT License as specified in LICENSE

PharmaJap committed:
feat(hook): add support for wildcard certificate verification (fixes #7)

Copyright (c) 2020 Justin Aplin for deSEC
Licensed under the MIT License as specified in LICENSE

PharmaJap committed:
fix(hook): respect minimum_ttl of domain (enables use on all domains)

Copyright (c) 2020 Justin Aplin for deSEC
Licensed under the MIT License as specified in LICENSE

PharmaJap committed:
feat(cleanup): add optional cleanup script (fixes #12)

Copyright (c) 2020 Justin Aplin for deSEC
Licensed under the MIT License as specified in LICENSE

peterthomassen commented on cleanup.sh:
```diff
@@ -0,0 +1,63 @@
+
+DEDYNAUTH=$(pwd)/.dedynauth
+
+if [ ! -f .dedynauth ]; then
```
Now that we have the variable, use `$DEDYNAUTH` here

PharmaJap replied:
True! Fixing.

peterthomassen commented on cleanup.sh:
```diff
@@ -0,0 +1,63 @@
+    exit 1
+fi
+
+source .dedynauth
```
use `$DEDYNAUTH`

PharmaJap replied:
True! Fixing.

peterthomassen commented on cleanup.sh:
```diff
@@ -0,0 +1,63 @@
+fi
+
+if [ -z "$CERTBOT_DOMAIN" ]; then
+    >&2 echo "It appears that you are not running this script through certbot (\$CERTBOT_DOMAIN is unset). Please call with: certbot --manual-auth-hook=$0"
```
Should be `--manual-cleanup-hook`, not `--manual-auth-hook`.

PharmaJap replied:
True! Fixing.

peterthomassen commented on cleanup.sh:
```diff
@@ -0,0 +1,63 @@
+    exit 3
+fi
+
+if [ -z "$CERTBOT_DOMAIN" ]; then
```
Variables passed along by certbot made me think of the following: It is possible to use the same script for both hooks, and discern within the script which is the call that is happening at the moment: [Only the cleanup script gets passed `$CERTBOT_AUTH_OUTPUT`](https://certbot.eff.org/docs/using.html?highlight=hook#pre-and-post-validation-hooks). If we could merge the two scripts, we would be able to save the first fourty lines of code duplication.

However, I don't think we need to do this now, just sharing my thoughts.

PharmaJap replied:
True! Actually, since we can `PUT` an empty records array to delete an rrset, we can reuse pretty much all of the code and just focus on manipulating `acme_records`. If you agree with https://github.com/desec-utils/certbot-hook/pull/13#discussion_r383116473 it's just a matter of testing for `$CERTBOT_AUTH_OUTPUT` to decide whether to add to or subtract from `acme_records` before republishing.

And now that I think about it, reusing the `host` lookup loop at the end to pause until we can verify the record has been deleted may solve some of the propagation issues with sequential certificates and make the `timeout_ttl` unnecessary... I will test.

peterthomassen replied:
Yes, that's great! I'm really impressive that you're willing to improve this script so much. Thanks a lot!

PharmaJap replied:
Looks like the cleanup verification loop did end up making the `timeout_ttl` unnecessary; dropping that commit. Pushing what I have now, but I'll work on another commit tonight that merges the two scripts.

peterthomassen commented on cleanup.sh:
```diff
@@ -0,0 +1,63 @@
+)
+
+# Delete ACME challenge(s)
+curl "${args[@]}" -X DELETE "https://desec.io/api/v1/domains/$DEDYN_NAME/rrsets/_acme-challenge$infix/TXT/"
```
This also deletes challenges that are there for other reasons which could cause errors in certain race condition situations (imagine using certbot from two different hosts at roughly the same time). As the cleanup script also gets passed `$CERTBOT_VALIDATION`, more granularity could be achieved by only removing the current challenge.

Having said that, following that approach seems a bit overengineered for this purpose. Maybe one day the API will provide record-level endpoints which would simplify this a lot. The client-side script is probably not the right place to do the logic.

PharmaJap replied:
Hmm... if we use an `shopt -s extglob`, we can basically duplicate the auth hook logic of adding/subtracting from `acme_records`. Something like:

```
if [ "$acme_records" = '"\"'"$CERTBOT_VALIDATION"'\""' ]; then
    acme_records=''
elif [[ $acme_records =~ "$CERTBOT_VALIDATION" ]]; then
    acme_records=${acme_records/@(',"\"'"$CERTBOT_VALIDATION"'\""'|'"\"'"$CERTBOT_VALIDATION"'\"",')/}
fi
```

And then `PUT` the result with the same `curl` call as the auth hook. That leaves the entire rest of the script basically identical... we'd be an if/then/else away from combining them entirely...

peterthomassen replied:
Awesome idea!

I suggest the following, which requires neither `@` nor the above `if`:

    acme_records=$acme_records,  # adds comma
    acme_records=${acme_records//*([[:space:]])'"\"'"$CERTBOT_VALIDATION"'\""'*([[:space:]])','*([[:space:]])/}  # removes all occurrences (double-slash)
    acme_records=${acme_records:0:-1}  # strip final comma (only if non-zero length)

The result can then be directly used in the `PUT` payload, like `"records": [$acme_records]`.

The `*([[:space:]])` patterns are there so that the matching does not depend on whitespace at the beginning of the string or before/after each comma. Because of this, `shopt -s extglob` is still required.

What do you think?

PharmaJap replied:
How are you figuring "only if non-zero length"? For me, I get:
```
"\"test1\""
"\"test1\"",

-bash: -1: substring expression < 0
```
Or do you mean test for non-zero length, then do the assignment?

peterthomassen replied:
I meant that a test would be required, yes

PharmaJap replied:
Works perfect!

peterthomassen commented on cleanup.sh:
```diff
@@ -0,0 +1,63 @@
+
+# Remove leading wildcard from infix, if present
+# *.foobar.dedyn.io gives "" while *.a.foobar.dedyn.io gives ".a"
+infix=${infix#.\*}
```
`infix=${infix#\*.}`

PharmaJap replied:
As in https://github.com/desec-utils/certbot-hook/pull/13#discussion_r383113996

peterthomassen replied:
yeah, my bad.

peterthomassen commented on hook.sh:
```diff
@@ -4,80 +4,109 @@
+
+DEDYNAUTH=$(pwd)/.dedynauth
+
 if [ ! -f .dedynauth ]; then
```
Now that we have the variable, use `$DEDYNAUTH` here

PharmaJap replied:
True! Fixing.

peterthomassen commented on hook.sh:
```diff
@@ -4,80 +4,109 @@
     exit 1
 fi
 
 source .dedynauth
```
Now that we have the variable, use `$DEDYNAUTH` here

PharmaJap replied:
True! Fixing.

peterthomassen reviewed (changes requested):
Very very nice work! Thanks a lot.

I found two small bugs which can be easily fixed. Looking forward to merging this!

peterthomassen commented:
@PharmaJap Do you need more guidance?

PharmaJap commented:
Sorry, I think my local copy is in decent shape; production hell has been following me home recently.

I'll see if I can push it out tonight :)

peterthomassen commented:
No rush! Just wanted to help in case you were waiting for something etc.

peterthomassen reviewed (approved):
Cool! Thank you so much. <3 I'll squash the last three commits and then merge.
